be/src/runtime/tablets_channel.cpp

```diff
@@ -196,77 +196,94 @@ void TabletsChannel::_close_wait(DeltaWriter* writer,
+    {
+        std::lock_guard<std::mutex> l(_lock);
+        if (_state == kFinished) {
+            // TabletsChannel is closed without LoadChannel's lock,
+            // therefore it's possible for reduce_mem_usage() to be called right after close()
+            return _close_status;
+        }
 
-    // Decide which writes should be flushed to reduce mem consumption.
-    // The main idea is to flush at least one third of the mem_limit.
-    // This is mainly to solve the following scenarios.
-    // Suppose there are N tablets in this TabletsChannel, and the mem limit is M.
-    // If the data is evenly distributed, when each tablet memory accumulates to M/N,
-    // the reduce memory operation will be triggered.
-    // At this time, the value of M/N may be much smaller than the value of `write_buffer_size`.
-    // If we flush all the tablets at this time, each tablet will generate a lot of small files.
-    // So here we only flush part of the tablet, and the next time the reduce memory operation is triggered,
-    // the tablet that has not been flushed before will accumulate more data, thereby reducing the number of flushes.
-
-    int64_t mem_to_flushed = mem_limit / 3;
-    int counter = 0;
-    int64_t sum = 0;
-    for (auto writer : writers) {
-        if (writer->memtable_consumption() <= 0) {
-            break;
+        // Sort the DeltaWriters by mem consumption in descend order.
+        std::vector<DeltaWriter*> writers;
+        for (auto& it : _tablet_writers) {
+            it.second->save_memtable_consumption_snapshot();
+            writers.push_back(it.second);
         }
-        ++counter;
-        sum += writer->memtable_consumption();
-        if (sum > mem_to_flushed) {
-            break;
+        std::sort(writers.begin(), writers.end(),
+                  [](const DeltaWriter* lhs, const DeltaWriter* rhs) {
+                      return lhs->get_memtable_consumption_snapshot() >
+                             rhs->get_memtable_consumption_snapshot();
+                  });
+
+        // Decide which writes should be flushed to reduce mem consumption.
+        // The main idea is to flush at least one third of the mem_limit.
+        // This is mainly to solve the following scenarios.
+        // Suppose there are N tablets in this TabletsChannel, and the mem limit is M.
+        // If the data is evenly distributed, when each tablet memory accumulates to M/N,
+        // the reduce memory operation will be triggered.
+        // At this time, the value of M/N may be much smaller than the value of `write_buffer_size`.
+        // If we flush all the tablets at this time, each tablet will generate a lot of small files.
+        // So here we only flush part of the tablet, and the next time the reduce memory operation is triggered,
+        // the tablet that has not been flushed before will accumulate more data, thereby reducing the number of flushes.
+
+        int64_t mem_to_flushed = mem_limit / 3;
```

Comment:
If a load channel has 20 tablet channels, a total of 9G memory is consumed, it seems unreasonable to choose the largest tablet channel to release 3G, because the largest tablet channel mem usage may only be 2G.

It seems that in `_find_largest_consumption_channel`, an algorithm is used to figure out how much memory each tablet channel is expected to release, and each tablet channel is traversed from large to small until the expected memory release reaches 1/3 of the load channel mem usage.

Reply:
I'm thinking, can we provide a method in `tablet channel` `delta writer` to get how much memory you expect to release + the delta writer that can release memtable (mem usage>50M (conf))

In the `handle_mem_exceed_limit` of `load channel`, flush these delta writers from large to small until the load channel mem usage * 1/3 is satisfied

Reply:
It seems reasonable to choose the largest tablet channel among the largest load channels, since there are usually only 1 or very few tablet channels.

Changing `mem_limit / 3` here to `tablet channel mem usage/3` seems to be fine.

Reply:
Offline discussed with xinyiZzz, some conclusion:
1. change the mem_limit / 3 to 'tabletsChannel's mem comsumption' /3
2. change the default value of soft limit from 80% to 50%, which can trigger the mem table flush more frequently, and still friendly to large load than old policy.